### PR TITLE
Support custom_metadata= argument in to_parquet

### LINKED ISSUE #7343: Define Column Metadata with to_parquet
When writing out to the `parquet` format, I want to define some column metadata.  Using pyarrow, I can write the metadata to the table using the following:

When writing out to parquet using pandas, I can create the Table using the pyarrow library, and once written, I can update the tables' metdata.  
```python
    metadata = table.schema.metadata
    metadata.update({b"geo": _encode_metadata(geo_metadata)})
```

With dask, can I somehow pass this information in above to the `to_parquet` class?

I see metadata files, but I want to define it on the table itself.

### PR DESCRIPTION
**Pull request by rjzamora**
Adds a `custom_metadata=` argument to `DataFrame.to_parquet`.  If a dictionary is passed in by the user, those key/values will be included in all footer metadata, and in the global _metadata file, along with the usual `b"pandas"` metadata.

~Note that this PR only adds `custom_metadata=` support for the "pyarrow" engine. @martindurant - Do you know if something similar can be accomplished in the "fastparquet" engine?  I have not investigated that end just yet.~ [EDIT: Both pyarrow and fastparquet engines are now supported.]

- [x] Closes #7343
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

**Comment by martindurant** [comment]
`fastparquet` itself does not support this, but it would be trivial to add.

File: dask/dataframe/io/parquet/arrow.py [hunk 1]
```diff
@@ -839,6 +840,10 @@ def write_partition(
             index_cols = []
 
         t = cls._pandas_to_arrow_table(df, preserve_index=preserve_index, schema=schema)
+        if custom_metadata:
+            _md = t.schema.metadata
+            _md.update(custom_metadata)
+            t = t.replace_schema_metadata(metadata=_md)
```
Comment (rjzamora) [inline dask/dataframe/io/parquet/arrow.py:846]: @jorisvandenbossche - I saw that this method is labeled as "Experimental," but it has been around for years now (as far as I can tell).  Is there any reason not to use it here?

**Comment by rjzamora** [comment]
>You may want to check for the pandas key, since that could render a dataset unreadable.

Can you clarify what you mean.  Are you saying that we shouldn't allow the user to modify the "pandas" metadata (i.e. add their own b"pandas" key)?

**Comment by martindurant** [comment]
In `FastParquetEngine.initialize_write` we create an `fmd` object, which is then used for the rest of the writes. We could insert metadata directly into this:
```
fmd.key_value_metadata.extend([
fastparquet.parquet_thrift.KeyValue(key=key, value=value)
for key, value in custom_metadata.items()
])
```

**Comment by martindurant** [comment]
> Can you clarify what you mean. Are you saying that we shouldn't allow the user to modify the "pandas" metadata (i.e. add their own b"pandas" key)?

Actually, the order matters - either we hand the user the ability to overwrite the pandas tag, or we would overwrite whatever they provide. In either case, I think a warning would be warranted.

**Comment by rjzamora** [comment]
>In FastParquetEngine.initialize_write we create an fmd object, which is then used for the rest of the writes. We could insert metadata directly into this

Nice! Good idea - I'll experiment with that :)

> Actually, the order matters

Just want to clarify... In the current solution, we are "updating" the key-value metadata with the user-specified metadata.  Therefore, we would be replacing an existing `b"pandas"` key if the user decided to pass in custom pandas metadata.  You are saying this is fine, but we want to add a warning, since changing the pandas metadata in the wrong way can render the data completely unreadable - correct?

**Comment by martindurant** [comment]
Correct. We *could* choose to error (you mustn't mess it up!) or ignore. Is there a genuine reason to want to update that tag? In our case, b"pandas" should always be present.

**Comment by jsignell** [comment]
Good to merge then?

**Comment by martindurant** [comment]
The one failure is indeed for parquet, so should be investigated

**Comment by rjzamora** [comment]
>The one failure is indeed for parquet, so should be investigated

I couldn't reproduce locally, and it doesn't seem that this PR was the cause (although I am not completely sure).

File: dask/dataframe/io/parquet/arrow.py (hunk 1 above)
Comment (jorisvandenbossche) [reply dask/dataframe/io/parquet/arrow.py:846]: I think that's fine to use (we should update the docstring)

Comment (jorisvandenbossche) [reply dask/dataframe/io/parquet/arrow.py:846]: -> https://issues.apache.org/jira/browse/ARROW-11980

File: dask/dataframe/io/tests/test_parquet.py [hunk 2]
```diff
@@ -3345,3 +3345,44 @@ def test_roundtrip_rename_columns(tmpdir, engine):
     df1.columns = ["d", "e", "f"]
 
     assert_eq(df1, ddf2.compute())
+
+
+def test_pyarrow_custom_metadata(tmpdir, engine):
```
Comment (jorisvandenbossche) [inline dask/dataframe/io/tests/test_parquet.py:3350]:
```suggestion
def test_custom_metadata(tmpdir, engine):
```

Comment (rjzamora) [reply dask/dataframe/io/tests/test_parquet.py]: Forgot about that - Thanks!

**Comment by jsignell** [comment]
Sorry this sat for so long @rjzamora

**Pull request merged**